docs/design.md

```diff
@@ -0,0 +1,156 @@
+# Hypper Design
+
+Hypper is a package manager built on Helm and inspired by other package managers
+like zypper.
+
+## How Hypper Differs From Helm
+
+Why build a package manager on a package manager? Helm is a package manager that
+tries to make few assumptions. For example, Custom Resource Definitions (CRD)s
+are a Kubernetes resource that extend the Kubernetes API. They can be used in
+various ways and there are pitfalls with them. Helm makes very few assumptions
+when it comes to them. Hypper is going to make some different assumptions which
+will result in different features.
+
+As differences between Hypper and Helm arise they will be documented along with
+the reasons for the difference. This is in an effort to provide clarity.
+
+When features from Hypper can be upstreamed into Helm they will be. In that way,
+Hypper can act as a testing ground for Helm. We do not expect that every feature
+will fit in upstream Helm.
+
+A second element in the difference is that Helm, as a package manager, does not
+handle environment deployment operations. Those are left to higher level tools
+like Helmfile (a push model like Ansible) and the Flux Helm Operator (a pull
+model like Chef). Helm expects higher level tools will use Helm and charts for
+operating in different environments.
+
+Hypper will move will make some assumptions that move it more towards the
```

Comment:
do we expect to have conf/env files like Helmfile to declare the full environment and the relationships between charts?

Reply:
We are not planning on it. Consider our current use cases...

1. Someone who wants to use a package manager. Do you want to maintain config files for zypper, homebrew, apt, or some other package manager? Do you ever touch config files for those?
2. As a library used in Rancher projects. There they will want to pass config around but not as files. And, Rancher is known for managing multiple clusters. We don't want to be in that business unless it's via fleet.

We want to keep the user experience in the realm of simple for end users. Whether they are using the API or CLI.

Reply:
> I don't understand the point 2: If the dependant package is only a helm library, there is no use of this, so there is indeed no need to pass file for the relationship, as the helm chart will declare itself the dependency.

Neither of these use cases are around managing the helm charts. They are both around how hypper itself is used. One case, the obvious one, is as a CLI that you use to install, update, and delete applications with.

The other use case... point 2... is when hypper itself is used as a library for other applications like rancher and fleet. For example, if fleet wanted to manage helm charts using hypper it would not call exec to a CLI. Instead it would import hypper as a software library and call Go APIs.

Does that make sense?

Reply:
:+1: